src/hotspot/share/opto/regmask.hpp

```diff
@@ -260,115 +548,299 @@ class RegMask {
     assert(valid_watermarks(), "sanity");
   }
 
   // AND 'rm' into 'this'
   void AND(const RegMask &rm) {
+    assert(_offset == rm._offset, "offset mismatch");
     assert(valid_watermarks() && rm.valid_watermarks(), "sanity");
-    // Do not evaluate words outside the current watermark range, as they are
-    // already zero and an &= would not change that
-    for (unsigned i = _lwm; i <= _hwm; i++) {
-      _RM_UP[i] &= rm._RM_UP[i];
+    _grow(rm._rm_size);
+    // Compute AND with all words from rm. Do not evaluate words outside the
+    // current watermark range, as they are already zero and an &= would not
+    // change that
+    for (unsigned int i = _lwm; i <= _hwm && i < rm._rm_size; i++) {
+      _rm_up(i) &= rm._rm_up(i);
+    }
+    // If rm is smaller than our high watermark and has the all-stack flag not
+    // set, we need to set all bits in the gap to 0.
+    if (!rm.is_AllStack() && _hwm > rm._rm_max()) {
+      _set_range(rm._rm_size, 0, _hwm - rm._rm_max());
+      _hwm = rm._rm_max();
+    }
+    // Narrow the watermarks if &rm spans a narrower range. Update after to
+    // ensure non-overlapping words are zeroed out. If rm has the all-stack
+    // flag set and is smaller than our high watermark, take care not to
+    // incorrectly lower the high watermark according to rm.
+    if (_lwm < rm._lwm) {
+      _lwm = rm._lwm;
+    }
+    if (_hwm > rm._hwm && !(rm.is_AllStack() && _hwm > rm._rm_max())) {
+      _hwm = rm._hwm;
     }
-    // Narrow the watermarks if &rm spans a narrower range.
-    // Update after to ensure non-overlapping words are zeroed out.
-    if (_lwm < rm._lwm) _lwm = rm._lwm;
-    if (_hwm > rm._hwm) _hwm = rm._hwm;
+    set_AllStack(is_AllStack() && rm.is_AllStack());
+    assert(valid_watermarks(), "sanity");
   }
 
-  // Subtract 'rm' from 'this'
+  // Subtract 'rm' from 'this'.
   void SUBTRACT(const RegMask &rm) {
+    assert(_offset == rm._offset, "offset mismatch");
     assert(valid_watermarks() && rm.valid_watermarks(), "sanity");
-    unsigned hwm = MIN2(_hwm, rm._hwm);
-    unsigned lwm = MAX2(_lwm, rm._lwm);
-    for (unsigned i = lwm; i <= hwm; i++) {
-      _RM_UP[i] &= ~rm._RM_UP[i];
+    _grow(rm._rm_size);
+    unsigned int hwm = MIN2(_hwm, rm._hwm);
+    unsigned int lwm = MAX2(_lwm, rm._lwm);
+    for (unsigned int i = lwm; i <= hwm; i++) {
+      _rm_up(i) &= ~rm._rm_up(i);
+    }
+    // If rm is smaller than our high watermark and has the all-stack flag set,
+    // we need to set all bits in the gap to 0.
+    if (rm.is_AllStack() && _hwm > rm._rm_max()) {
+      _set_range(rm.rm_size(), 0, _hwm - rm._rm_max());
+      _hwm = rm._rm_max();
+    }
+    set_AllStack(is_AllStack() && !rm.is_AllStack());
+    _trim_watermarks();
+    assert(valid_watermarks(), "sanity");
+  }
+
+  // Subtract 'rm' from 'this', but ignore everything in 'rm' that does not
+  // overlap with us and to not modify our all-stack flag. Supports masks of
```

Comment:
A little confused about the grammar in `and to not modify our all-stack flag`...

Reply:
Thanks, typo. Fixed "to" -> "do"